src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/HandshakeProtocol.cs

```diff
@@ -70,19 +70,38 @@ public static HandshakeResponseMessage ParseResponseMessage(ReadOnlyMemory<byte>
         {
             using (var reader = CreateJsonTextReader(payload))
             {
-                var token = JToken.ReadFrom(reader);
-                var handshakeJObject = JsonUtils.GetObject(token);
+                JsonUtils.CheckRead(reader);
+                JsonUtils.EnsureObjectStart(reader);
 
-                // a handshake response does not have a type
-                // check the incoming message was not any other type of message
-                var type = JsonUtils.GetOptionalProperty<string>(handshakeJObject, TypePropertyName);
-                if (!string.IsNullOrEmpty(type))
+                string error = null;
+
+                var completed = false;
+                do
                 {
-                    throw new InvalidOperationException("Handshake response should not have a 'type' value.");
+                    switch (reader.TokenType)
+                    {
+                        case JsonToken.PropertyName:
+                            string memberName = reader.Value.ToString();
+
+                            switch (memberName)
+                            {
+                                case TypePropertyName:
+                                    // a handshake response does not have a type
+                                    // check the incoming message was not any other type of message
+                                    throw new InvalidOperationException("Handshake response should not have a 'type' value.");
+                                case ErrorPropertyName:
+                                    error = JsonUtils.ReadAsString(reader, ErrorPropertyName);
+                                    break;
+                            }
+                            break;
+                        case JsonToken.EndObject:
+                            completed = true;
+                            break;
+                    }
                 }
+                while (!completed && JsonUtils.CheckRead(reader));
 
-                var error = JsonUtils.GetOptionalProperty<string>(handshakeJObject, ErrorPropertyName);
-                return new HandshakeResponseMessage(error);
+                return (!string.IsNullOrEmpty(error)) ? new HandshakeResponseMessage(error) : HandshakeResponseMessage.Empty;
```

Comment:
Is a handshake response with an empty error string in the JSON considered successful?

Reply:
Good question :). I don't have a clear opinion either way, as long as we make a clear decision. The simplest seems to be to say that if the `error` property is present (even if it's `null`, `""`, or even an invalid type) the response is an error. In fact, sending a negotiation response with an invalid `error` property would essentially mean that the server is not behaving properly and the client can't trust it anyway, so it should fail.